onnxruntime/core/providers/nnapi/nnapi_builtin/builders/impl/base_op_builder.cc

```diff
@@ -51,8 +53,12 @@ Status BaseOpBuilder::AddToModelBuilder(ModelBuilder& model_builder, const NodeU
       model_builder.GetEffectiveFeatureLevel(),
       model_builder.UseNCHW(),
   };
-  ORT_RETURN_IF_NOT(IsOpSupported(model_builder.GetInitializerTensors(), node_unit, params),
-                    "Unsupported operator ", node_unit.OpType());
+
+  // We checked supported in IExecutionProvider::GetCapability.
+  // Checking again in AddToModelBuilder which is called in IExecutionProvider::Compile is redundant.
+  // ORT_RETURN_IF_NOT(IsOpSupported(model_builder.GetGraphViewer(), node_unit, params),
+  //                  "Unsupported operator ", node_unit.OpType());
```

Comment:
I don't see a reason for calling IsOpSupported again creating the model in IExecutionProvider::Compile. Compilation happens with the set of nodes returned by GetCapability, and if there's a layout change it's the set of nodes returned by the call to GetCapability post-layout change. nothiing I'm aware of changes the nodes between these two steps so the call seems redundant.

Reply:
is it part of the contract between ORT and an EP that only nodes that the EP reports it can handle (with `GetCapability()`) will be given to it via `Compile()`?

that was my assumption, although I didn't find it explicitly documented.

Reply:
That's how it works currently but I don't believe that's a guarantee. 

If there was more advanced partitioning ORT _could_ decide a lower priority EP is assigned some/all of those nodes. e.g. if it decided it wanted to allow only part of the model to execute using GPU to save GPU memory, a higher priority GPU based EP could request nodes, but ORT could decide a lower priority CPU based EP will be assigned them based on memory usage characteristics.

But AFAIK the only thing that changes the actual nodes during partitioning is layout transformation, that only affects nodes that are going to be assigned to the EP, and GetCapability it called again after that to validate.

Reply:
but can an EP assume that it will only receive nodes from `Compile()` which it has previously said it could handle with `GetCapability()`?

if that's the case, then we shouldn't need this `IsOpSupported` call.

Reply:
That is the case. What is returned by GetCapability is the superset of all nodes that Compile may be called with.